docs/standard/serialization/snippets/system-text-json-how-to/csharp/RoundtripDataTable.cs

```diff
@@ -0,0 +1,246 @@
+    }
+
+    public static class Program
+    {
+        public static List<WeatherForecast> weatherForecasts = new List<WeatherForecast>()
+        {
+            new WeatherForecast { Date = new DateTime(2019, 8, 1), TemperatureCelsius = 25, Summary = "Hot" },
+            new WeatherForecast { Date = new DateTime(2019, 8, 2), TemperatureCelsius = 20, Summary = "Warm" },
+            new WeatherForecast { Date = new DateTime(2019, 8, 3), TemperatureCelsius = 10, Summary = "Cold" }
+        };
+
+        public static void Main()
+        {
+            // Serialize a List<T> object and display the JSON
+            string jsonString = JsonSerializer.Serialize(
+                weatherForecasts,
+                new JsonSerializerOptions { WriteIndented = true });
+            Console.WriteLine(jsonString);
+
+            // Deserialize to a DataTable
+            var deserializeOptions = new JsonSerializerOptions();
+            deserializeOptions.Converters.Add(new DataTableJsonConverter());
+            var weatherForecastTable = JsonSerializer.Deserialize<DataTable>(jsonString, deserializeOptions);
+
+            // Display the DataTable contents
+            foreach (DataRow row in weatherForecastTable.Rows)
+            {
+                for (int i = 0; i < weatherForecastTable.Columns.Count; i++)
+                {
+                    Console.Write($"{row[i]} ");
+                }
+                Console.WriteLine();
+            }
+
+            // Serialize the DataTable and display the JSON
+            var serializeOptions = new JsonSerializerOptions();
+            serializeOptions.WriteIndented = true;
+            serializeOptions.Converters.Add(new DataTableJsonConverter());
+            jsonString = JsonSerializer.Serialize(weatherForecasts, serializeOptions);
+            Console.WriteLine(jsonString);
+        }
+    }
+
+    public class DataTableJsonConverter : JsonConverter<DataTable>
+    {
+        public override DataTable Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            using var jsonDoc = JsonDocument.ParseValue(ref reader);
+            var rootElement = jsonDoc.RootElement;
+            var dataTable = rootElement.JsonElementToDataTable();
+            return dataTable;
+        }
+
+        public override void Write(Utf8JsonWriter jsonWriter, DataTable value, JsonSerializerOptions options)
+        {
+            jsonWriter.WriteStartArray();
+            foreach (DataRow dr in value.Rows)
+            {
+                jsonWriter.WriteStartObject();
+                foreach (DataColumn col in value.Columns)
+                {
+                    var key = col.ColumnName.Trim();
+                    var valueString = dr[col].ToString();
+                    switch (col.DataType.FullName)
+                    {
+                        case "System.Guid":
+                            jsonWriter.WriteString(key, valueString);
```

Comment:
Can we do a test to see if leveraging `Utf8JsonWriter` support for all these primitive types is more performant? Similar to the work you did with our guidance on reusing `JsonSerializerOptions` instances, we can use `StopWatch` to check elapsed time (and [Process.PrivateMemorySize64](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.privatememorysize64?view=net-5.0#System_Diagnostics_Process_PrivateMemorySize64)) to check allocations. We'd do this here and for all types below.

```suggestion
                            jsonWriter.WriteString(key, (Guid)dr[col]);
```

Rather than convert to a string, we unbox the primitive type and use the native writer support directly.

Reply:
I tried Utf8JsonWriter support for primitive types and tested with 1,000,000 rows in the table -- result was lower allocations (MemorySize64 1,306M vs. 1,316M) but more elapsed time (952 vs. 877). I left the new code in place.